### PR TITLE
feat(web): per-queue size & latency historical gauges (#124, #126)

### LINKED ISSUE #124: [parity/Ent] Add per-queue size and latency historical time-series (Ent Historical tab)
## What
Sidekiq Enterprise §5.2 emits `sidekiq.queue.size` and `sidekiq.queue.latency` gauges tagged `queue:<name>`; the Historical tab (§7 line 712) renders per-queue gauges over a recent window. wurk captures only per-job-class series.

Reference: docs/target/sidekiq-ent.md §5.2, §7 routes table line 712. Wiki: https://github.com/sidekiq/sidekiq/wiki/Ent-Historical-Metrics

## Current state
- lib/wurk/metrics/query.rb aggregates per-job-class buckets only.
- lib/wurk/web/enterprise.rb Historical module has no per-queue endpoint.
- No queue-level time-series captured.

## Acceptance criteria
- [ ] Periodic capture of each queue's size and head-of-line latency into time-buckets (or via the §5 snapshotter's default emitter writing to Redis).
- [ ] A query API returning per-queue size/latency series over a window.
- [ ] Historical tab / Metrics page renders per-queue gauges over the recent window.
- [ ] Tests covering capture + query for at least two queues.

---
**Tier:** enterprise · **Audit status:** missing · **Reference:** docs/target/sidekiq-ent.md §5.2 (sidekiq.queue.size / sidekiq.queue.latency) and §7 routes table line 712; wiki https://github.com/sidekiq/sidekiq/wiki/Ent-Historical-Metrics

_Filed from the automated Sidekiq parity audit (milestone: **Sidekiq parity**). Verified as a real gap via adversarial re-check:_ Gap stands. wurk has only per-job-CLASS time-series: lib/wurk/metrics/query.rb (for_job/top_jobs/history all keyed on class name, never queue), backed by lib/wurk/metrics/history.rb (j|YYMMDD|H:M HASH, <klass>|p|f|ms fields) and lib/wurk/metrics/rollup.rb:34-149 (jr|<bucket>|<epoch> HASH summing p/f/ms across classes — no queue dimension). No per-queue size/latency series is persisted or queryable. lib/wurk/heartbeat.rb:185-196 emit_queue_sizes pushes sidekiq.queue.size per queue to an external dogstatsd UDP socket (fire-and-forget via lib/wurk/metrics/statsd.rb:57-68), but (a) never emits sid

### LINKED ISSUE #126: [parity/Ent] Add per-queue size/latency gauge charts to the Metrics/Historical dashboard
## What
The Ent Historical tab (§7 line 712) shows per-queue and global gauges over a recent window. wurk's Metrics page shows job throughput/failure line charts and a top-jobs bar chart, but no per-queue size/latency gauge charts.

Reference: docs/target/sidekiq-ent.md §5.2 (queue.size/queue.latency), §7 routes table line 712. Depends on the per-queue time-series capture issue.

## Current state
- frontend/src/pages/Metrics.tsx:62-208 renders cluster-total LineChart + top-jobs BarChart only.
- No per-queue gauge series exists to chart.

## Acceptance criteria
- [ ] Once per-queue size/latency series exist, add chart(s) showing each queue's size and latency over the selected window.
- [ ] Window/bucket selector applies to the queue charts.
- [ ] Empty-state handling when no queue history yet.
- [ ] Render test for the new charts.

---
**Tier:** enterprise · **Audit status:** partial · **Reference:** docs/target/sidekiq-ent.md §5.3 and §7 routes table line 712 (Historical tab: per-queue and global gauges); wiki https://github.com/sidekiq/sidekiq/wiki/Ent-Historical-Metrics

_Filed from the automated Sidekiq parity audit (milestone: **Sidekiq parity**). Verified as a real gap via adversarial re-check:_ Gap is REAL. The Ent Historical tab's per-queue size/latency gauges-over-time are not implemented, and no intentional divergence is documented.

What wurk has (job-class throughput/failures only):
- frontend/src/pages/Metrics.tsx:62-118 HistoryCharts renders ONLY cluster-total processed/failed Lines from /wurk/api/history/:bucket; lines 178-247 a top-jobs BarChart. No per-queue dimension anywhere.
- lib/wurk/metrics/rollup.rb:30-39,107-116 rolls up per-JOB-CLASS p/f/ms into jr|1m|5m|1h cluster-total buckets — summed over classes, NOT per-queue, and no size/latency captured.
- lib/wurk/metrics/

### PR DESCRIPTION
**Pull request by dinndev2**
Closes #124. Closes #126.

Sidekiq Enterprise §5.2 emits per-queue `sidekiq.queue.size` / `sidekiq.queue.latency` gauges, and the Historical tab (§7) charts them over a recent window. wurk captured only per-job-**class** throughput — no per-queue dimension was persisted or queryable. This adds it end to end. #126 (charts) depends on #124 (the series), so they ship together.

## Backend (#124 — the data)
- **`Metrics::QueueRollup`** — a leader-gated sampler thread (mirrors `Metrics::Rollup`). Each tick snapshots every live queue's depth (`LLEN`) + head-of-line latency (tail of the LIST, exactly like `Queue#latency`) into gauge buckets `qm|{1m,5m,1h}|<epoch>` with `<queue>|sz` / `<queue>|lt` fields, at all three resolutions, with 24h/7d/30d retention. Size/latency are **gauges**, so a tick samples "now" and overwrites the current bucket (last-value downsample) rather than summing counters.
- **`Query.queue_history`** + **`Enterprise::Historical.queue_history`** readers — gap-filled, window-clamped, `MAX_QUEUE_SERIES`-capped; mirror the existing `history` reader.
- **`GET /wurk/api/queue-history/:bucket?window=&queue=`** + a `queue_history_series` serializer.
- Launcher builds/starts/terminates the sampler alongside the metrics rollup (leader-gated, so only one process writes).

## Frontend (#126 — the charts)
- **`QueueGaugeCharts`** on the Metrics page: stacked **Depth** + **Latency** line charts, one line per queue, the same `24h·1m / 7d·5m / 30d·1h` range selector as the throughput chart, and an empty state.

## Acceptance criteria
**#124**
- [x] Periodic capture of each queue's size + head-of-line latency into time-buckets.
- [x] A query API returning per-queue size/latency series over a window.
- [x] Metrics page renders per-queue gauges over the recent window.
- [x] Tests covering capture + query for ≥2 queues.

**#126**
- [x] Per-queue size & latency charts over the selected window.
- [x] Window/bucket selector applies to the queue charts.
- [x] Empty-state handling when no queue history yet.
- [x] Render test for the new charts.

## Tests
- `test/unit/metrics_queue_rollup_test.rb` — capture for 2+ queues, head-of-line latency, per-bucket TTLs, leader gating, thread loop.
- `test/unit/metrics_query_test.rb` — `queue_history` series, gap-fill, window clamp, bad bucket, live-queue-set path, queue cap.
- `test/engine/api_endpoints_test.rb` — endpoint returns per-queue gauge series, single-queue filter, bad bucket → 400.
- **First frontend test runner** (vitest + testing-library): `frontend/src/pages/Metrics.test.tsx` asserts the empty state and one `<Line>` per queue across both charts.
- Full suite **2165 runs, 0 failures**, coverage gate, `test:parity`, `test:ecosystem`, and `tsc -b && vite build` all green locally.

**Wire-compat:** only adds new `qm|…` keys; no existing Redis key or job JSON changed.

## How to test in prod
Per-queue gauges are captured automatically once a worker boots (the leader samples every ~minute). Enqueue some jobs to spread depth/latency across a few queues:

```ruby
1000.times { SomeJob.set(queue: "default").perform_async }
50.times   { SomeJob.set(queue: "critical").perform_async }
```

Open the dashboard → **Metrics** → **Queue Size & Latency**. After a couple of minutes you'll see a line per queue in both the Depth and Latency charts; the `24h·1m / 7d·5m / 30d·1h` selector switches the window. New JSON: `GET /wurk/api/queue-history/1m?window=24h`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Per-queue historical depth and latency charts with time-window selection and optional queue filtering; a new JSON API provides the chart data and periodic sampling runs server-side.

* **Tests**
  * Extensive unit and integration tests covering queue metric sampling, retention/edge cases, API responses (including errors/filters), and frontend chart/error/empty states.

* **Chores**
  * Frontend test tooling and scripts added to run the new test suite.

### CODE REVIEW

File: frontend/src/pages/Metrics.tsx [hunk 1]
```diff
@@ -119,6 +164,95 @@ function HistoryCharts() {
...
+function QueueGaugeCharts() {
+  const [range, setRange] = useState(0);
+  const { bucket, window } = HISTORY_RANGES[range];
+
+  const { data, isLoading, isError } = useQuery<QueueHistoryResponse>({
+    queryKey: ['queue-history', bucket, window],
+    queryFn: () =>
+      fetch(`/wurk/api/queue-history/${bucket}?window=${window}`).then(
+        (r) => r.json() as Promise<QueueHistoryResponse>
+      ),
```
[the remark this answers is not in the document]
Comment (dinndev2) [reply frontend/src/pages/Metrics.tsx]: Fixed in 1eff763 — the queue-history queryFn now throws on a non-OK response (parsing the `error` payload for the message), so `isError` fires instead of an error body falling through to the empty state. Added a frontend error-path test.

File: lib/wurk/metrics/queue_rollup.rb [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        end
+      end
+
+      def head_latency(lrange_result, now_ms)
+        payload = lrange_result.is_a?(::Array) ? lrange_result.first : lrange_result
+        return 0.0 if payload.nil?
+
+        Wurk::JobRecord.latency_from(Wurk.load_json(payload)['enqueued_at'], now_ms).round(2)
+      rescue ::JSON::ParserError
+        0.0
```
[the remark this answers is not in the document]
Comment (dinndev2) [reply lib/wurk/metrics/queue_rollup.rb:141]: Fixed in 1eff763 — `head_latency` now guards a non-Hash parse result and rescues `TypeError`/`ArgumentError` in addition to `JSON::ParserError`, so one malformed tail payload yields 0.0 for that queue instead of skipping the whole tick. Added a malformed-payload test.

File: test/unit/metrics_query_test.rb [hunk 3]
```diff
@@ -222,4 +222,71 @@ def test_pipeline_hgetall_empty_keys_returns_empty
...
+  end
+
+  def test_queue_history_caps_queue_count
+    names = (0...30).map { |i| format('Qcap%<s>s_%<i>02d', s: @suffix, i: i) }
+    Wurk.redis { |c| c.call('SADD', 'queues', *names) }
+
+    series = Wurk::Metrics::Query.queue_history('1m', 120, now: @now)
+
+    assert_operator series.size, :<=, Wurk::Metrics::Query::MAX_QUEUE_SERIES
+  end
```
[the remark this answers is not in the document]
Comment (dinndev2) [reply test/unit/metrics_query_test.rb:285]: Fixed in 1eff763 — teardown now SREMs this suite's members from the shared `queues` set. (The base RedisNamespace teardown also FLUSHDBs between tests, but the explicit SREM makes the intent clear and removes any ordering risk.)

**Comment by dinndev2** [comment]
@coderabbitai review

**Comment by dinndev2** [comment]
Addressed both nitpicks in aa18085 — merged the multi-assertion tests into single hash comparisons and let autocorrect fix the spacing; extracted helpers so the changed methods are back under RuboCop's AbcSize threshold. @coderabbitai review
